File: matlab/dmp_lib/DMP/@DMP_pos/DMP_pos.m
%% DMP_pos class
%  For encoding Cartesian Orientation.
%


classdef DMP_pos < matlab.mixin.Copyable

    methods  (Access = public)
        %% DMP_pos constructor.
        %  @param[in] N_kernels: 3x3 matrix where each row contains the kernels for position,
        %                       velocity and acceleration and each row for x, y and z coordinate.
        %                       If 'N_kernels' is a column vector then every coordinate will have the
        %                       same number of kernels.
        %  @param[in] a_z: Parameter 'a_z' relating to the spring-damper system.
        %  @param[in] b_z: Parameter 'b_z' relating to the spring-damper system.
        %  @param[in] can_clock_ptr: Pointer to a DMP canonical system object.
        %  @param[in] shape_attr_gating_ptr: Pointer to a DMP gating function object.
        %
        function this = DMP_pos(dmp_type, N_kernels, a_z, b_z, can_clock_ptr, shape_attr_gating_ptr)
            
            if (nargin < 5), can_clock_ptr = CanonicalClock(); end
            if (nargin < 6), shape_attr_gating_ptr=SigmoidGatingFunction(1.0, 0.5); end
            
            if (isscalar(N_kernels)), N_kernels = ones(3,1)*N_kernels; end
            if (isscalar(a_z)), a_z = ones(3,1)*a_z; end
            if (isscalar(b_z)), b_z = ones(3,1)*b_z; end

            this.can_clock_ptr = can_clock_ptr;
            this.shape_attr_gating_ptr = shape_attr_gating_ptr;

            for i=1:3
                if (dmp_type == DMP_TYPE.STD)
                    this.dmp{i} = DMP(N_kernels(i), a_z(i), b_z(i), can_clock_ptr, shape_attr_gating_ptr);
                elseif (dmp_type == DMP_TYPE.BIO)
                    this.dmp{i} = DMP_bio(N_kernels(i), a_z(i), b_z(i), can_clock_ptr, shape_attr_gating_ptr);
                else
                    error('[DMP_pos::DMP_pos]: Unsopported DMP type!');
                end
            end

        end
        
        
        %% Trains the DMP_pos.
        %  @param[in] train_method: The training method (see dmp_::TRAIN_METHOD enum).
        %  @param[in] Time: 1xN row vector with the timestamps of the training data points.
        %  @param[in] Pd_data: 3xN matrix with desired position at each column for each timestep.
        %  @param[in] dPd_data: 3xN matrix with desired velocity at each column for each timestep.
        %  @param[in] ddPd_data: 3xN matrix with desired acceleration at each column for each timestep.
        %  @param[in] train_error: Optinal pointer to return the training error as norm(F-Fd)/n_data.
        function train_err = train(this, train_method, Time, Pd_data, dPd_data, ddPd_data)
            
            tau = Time(end);
            this.setTau(tau);

            if (nargout > 0)
                train_err = zeros(3,1);
                for i=1:3
                    train_err(i) = this.dmp{i}.train(train_method, Time, Pd_data(i,:), dPd_data(i,:), ddPd_data(i,:));
                end
            else
                for i=1:3
                    this.dmp{i}.train(train_method, Time, Pd_data(i,:), dPd_data(i,:), ddPd_data(i,:));
                end
            end

        end
        
        %% Calculates the derivatives of the DMP states. The derivatives can then be
        %% retrieved with 'getXdot', 'getYdot' and 'getZdot'.
        %  @param[in] x: phase variable.
        %  @param[in] Y: 'y' state of the DMP.
        %  @param[in] Z: 'z' state of the DMP.
        %  @param[in] Y0: Initial position.
        %  @param[in] Yg: Goal position.
        %  @param[in] Y_c: Coupling term for the dynamical equation of the 'y' state.
        %  @param[in] Z_c: Coupling term for the dynamical equation of the 'z' state.
        function update(this, x, Y, Z, Y0, Yg, Y_c, Z_c)
            
            if (nargin < 7), Y_c=zeros(3,1); end
            if (nargin < 8), Z_c=zeros(3,1); end
            
            for i=1:3, this.dmp{i}.update(x, Y(i), Z(i), Y0(i), Yg(i), Y_c(i), Z_c(i)); end
            
            this.dZ = zeros(3,1);
            this.dY = zeros(3,1);
            for i=1:3
                this.dY(i) = this.dmp{i}.getYdot();
                this.dZ(i) = this.dmp{i}.getZdot();    
            end
            this.dx = this.phaseDot(x);

        end
        
        function ddY = calcYddot(this, x, Y, dY, Y0, Yg, tau_dot, Yc, Zc, Yc_dot)
            
            if (nargin < 7), tau_dot = 0; end
            if (nargin < 8), Yc = zeros(3,1); end
            if (nargin < 9), Zc = zeros(3,1); end
            if (nargin < 10), Yc_dot = zeros(3,1); end

            ddY = zeros(3,1);
            for i=1:3
               ddY(i) = this.dmp{i}.calcYddot(x, Y(i), dY(i), Y0(i), Yg(i), tau_dot, Yc(i), Zc(i), Yc_dot(i)); 
            end
            
        end

        
        %% Returns the acceleration for the given input state defined by the timestamp,
        %  the orientation, the angular velocity and acceleration, the initial and target orientation
        %  and an optinal coupling term.
        %  @param[in] x: phase variable.
        %  @param[in] P: Current position.
        %  @param[in] dP: Current velocity.
        %  @param[in] P0: Initial position.
        %  @param[in] Pg: Goal position.
        %  @param[in] Z_c: Coupling term. (optional, default=arma::vec().zeros(3))
        %  @return ddP: Acceleration.
        %
        function Yddot = getYddot(this, tau_dot, Yc_dot)
            
            if (nargin < 2), tau_dot = 0; end
            if (nargin < 3), Yc_dot = zeros(3,1); end
            
            Yddot = zeros(3,1);
            for i=1:3
                Yddot(i) = (this.dmp{i}.getZdot() + Yc_dot - tau_dot*this.dmp{i}.getYdot()) / this.dmp{i}.getTau(); 
            end
            
        end
        

        %% Returns the time scaling factor.
        %  @return: The time scaling factor.
        %
        function tau = getTau(this)

            tau = this.can_clock_ptr.getTau();

        end
        
        
        %% Sets the time scaling factor.
        %  @param[in] tau: The time scaling factor.
        %
        function setTau(this, tau)

            this.can_clock_ptr.setTau(tau);

        end
        
        
        %% Returns the phase variable corresponding to the given time instant.
        %  @param[in] t: The time instant.
        %  @return: The phase variable for time 't'.
        %
        function x = phase(this, t)
            
            x = this.can_clock_ptr.getPhase(t);

        end
        
        
        %% Returns the derivative of the phase variable.
        %  @param[in] x: The phase variable.
        %  @return: The derivative of the phase variable.
        %
        function dx = phaseDot(this, x)
            
            dx = this.can_clock_ptr.getPhaseDot(x);

        end
        
        function dx = getXdot(this), dx=this.dx; end
        function dY = getYdot(this), dY=this.dY; end
        function dZ = getZdot(this), dZ=this.dZ; end
        
        function can_clock_ptr = getCanClockPtr(this), can_clock_ptr = this.can_clock_ptr; end
        
    end
    
    
    properties  (Access = private)
        
        dmp % 3x1 vector of DMP producing the desired trajectory for each x, y and z dimensions.

        can_clock_ptr % handle (pointer) to the canonical clock
        shape_attr_gating_ptr % pointer to gating function for the shape attractor
        
        dZ
        dY
        dx
    end
    
end

File: c++/src/dmp_lib/src/DMP/DMP_pos.cpp
#include <dmp_lib/DMP/DMP_pos.h>
#include <dmp_lib/DMP/DMP.h>
#include <dmp_lib/DMP/DMP_bio.h>

#include <dmp_lib/GatingFunction/LinGatingFunction.h>
#include <dmp_lib/GatingFunction/ExpGatingFunction.h>
#include <dmp_lib/GatingFunction/SigmoidGatingFunction.h>

#include <dmp_lib/io/io.h>

namespace as64_
{

namespace dmp_
{

DMP_pos::DMP_pos(dmp_::TYPE dmp_type, const arma::uvec &N_kernels, const arma::vec &a_z, const arma::vec &b_z,
                 std::shared_ptr<CanonicalClock> can_clock_ptr, std::shared_ptr<GatingFunction> shape_attr_gating_ptr)
{

  if (!can_clock_ptr) can_clock_ptr.reset(new CanonicalClock());
  if (!shape_attr_gating_ptr) shape_attr_gating_ptr.reset(new SigmoidGatingFunction(1.0, 0.5));

  this->can_clock_ptr = can_clock_ptr;
  this->shape_attr_gating_ptr = shape_attr_gating_ptr;

  dmp.resize(3);
  for (int i=0; i<dmp.size(); i++)
  {
    switch (dmp_type)
    {
      case dmp_::TYPE::STD:
        this->dmp[i].reset(new DMP(N_kernels(i), a_z(i), b_z(i), can_clock_ptr, shape_attr_gating_ptr));
        break;
      case dmp_::TYPE::BIO:
        this->dmp[i].reset(new DMP_bio(N_kernels(i), a_z(i), b_z(i), can_clock_ptr, shape_attr_gating_ptr));
        break;
      default:
        throw std::runtime_error("[DMP_pos::DMP_pos]: Unsupported DMP type!");
    }
  }

}


void DMP_pos::train(dmp_::TrainMethod train_method, const arma::rowvec &Time, const arma::mat &Pd_data,
    const arma::mat &dPd_data, const arma::mat &ddPd_data, arma::vec *train_err)
{
  int i_end = Time.size()-1;
  double tau = Time(i_end);
  this->setTau(tau);

  int n_dim = this->dmp.size();

  if (train_err)
  {
    train_err->resize(n_dim);
    for (int i=0; i<n_dim; i++)
      this->dmp[i]->train(train_method, Time, Pd_data.row(i), dPd_data.row(i), ddPd_data.row(i), &(train_err->at(i)));
  }
  else
  {
    for (int i=0; i<n_dim; i++)
      this->dmp[i]->train(train_method, Time, Pd_data.row(i), dPd_data.row(i), ddPd_data.row(i));
  }

}


void DMP_pos::update(double x, const arma::vec &Y, const arma::vec &Z, const arma::vec &Yg, const arma::vec &Y_c, const arma::vec &Z_c)
{
  int n_dim = this->dmp.size();

  for (int i=0; i<n_dim; i++) this->dmp[i]->update(x, Y(i), Z(i), Yg(i), Y_c(i), Z_c(i));

  this->dZ.resize(n_dim);
  this->dY.resize(n_dim);
  for (int i=0; i<n_dim; i++)
  {
    this->dY(i) = this->dmp[i]->getYdot();
    this->dZ(i) = this->dmp[i]->getZdot();
  }

  this->dx = this->phaseDot(x);

}


arma::vec DMP_pos::calcYddot(double x, const arma::vec &Y, const arma::vec &dY, const arma::vec &Yg,
                             double tau_dot, const arma::vec &Yc, const arma::vec &Zc, const arma::vec &Yc_dot)
{
  int n_dim = this->dmp.size();
  arma::vec ddY(n_dim);
  for (int i=0; i<n_dim; i++)
    ddY(i) = this->dmp[i]->calcYddot(x, Y(i), dY(i), Yg(i), tau_dot, Yc(i), Zc(i), Yc_dot(i));

  return ddY;
}


arma::vec DMP_pos::getYddot(double tau_dot, const arma::vec &Yc_dot) const
{
  return (this->getZdot() + Yc_dot - tau_dot*this->getYdot()) / this->getTau();
}


void DMP_pos::setY0(const arma::vec &Y0)
{
  for (int i=0; i<dmp.size(); i++) this->dmp[i]->setY0(Y0(i));
}


arma::vec DMP_pos::getY0() const
{
  arma::vec Y0(dmp.size());
  for (int i=0; i<Y0.size(); i++) Y0(i) = this->dmp[i]->getY0();
  return Y0;
}


void DMP_pos::setTau(double tau)
{
  this->can_clock_ptr->setTau(tau);
}


double DMP_pos::getTau() const
{
  return this->can_clock_ptr->getTau();
}


double DMP_pos::phase(double t) const
{
  return this->can_clock_ptr->getPhase(t);
}


double DMP_pos::phaseDot(double x) const
{
  return this->can_clock_ptr->getPhaseDot(x);
}


void DMP_pos::exportToFile(std::ostream &out) const
{
  for (int i=0; i<dmp.size(); i++) dmp[i]->exportToFile(out);
  can_clock_ptr->exportToFile(out);
  shape_attr_gating_ptr->exportToFile(out);
}

std::shared_ptr<DMP_pos> DMP_pos::importFromFile(std::istream &in)
{
  std::shared_ptr<DMP_pos> dmp_p(new DMP_pos(dmp_::STD, {30,30,30}, {20,20,20}, {5,5,5}));

  int N_dmps = dmp_p->dmp.size();
  for (int i=0; i<N_dmps; i++) dmp_p->dmp[i] = DMP_::importFromFile(in);
  dmp_p->can_clock_ptr = CanonicalClock::importFromFile(in);
  dmp_p->shape_attr_gating_ptr = GatingFunction::importFromFile(in);

  for (int i=0; i<N_dmps; i++)
  {
    dmp_p->dmp[i]->can_clock_ptr = dmp_p->can_clock_ptr;
    dmp_p->dmp[i]->shape_attr_gating_ptr = dmp_p->shape_attr_gating_ptr;
  }

  return dmp_p;
}

arma::mat DMP_pos::getAcellPartDev_g_tau(double t, const arma::vec &Y, const arma::vec &dY, const arma::vec &Y0, double x, const arma::vec &Yg, double tau) const
{

  int n_dim = dmp.size();
  arma::mat J(n_dim, n_dim+1);

  int j_end = n_dim;
  for (int i=0; i<n_dim; i++)
  {
    arma::vec C = dmp[i]->getAcellPartDev_g_tau(t, Y(i), dY(i), Y0(i), x, Yg(i), tau);
    J(i,i) = C(0);
    J(i,j_end) = C(1);
  }

  return J;
}

} // namespace dmp_

} // namespace as64_
